src/System/extract_state.rs

```diff
@@ -1,66 +1,115 @@
 use edbm::zones::OwnedFederation;
 
+use crate::component::Declarations;
 use crate::extract_system_rep::SystemRecipe;
 use crate::EdgeEval::constraint_applyer::apply_constraints_to_state;
 use crate::ModelObjects::component::State;
-use crate::ModelObjects::representations::{BoolExpression, QueryExpression};
+use crate::ModelObjects::representations::{BoolExpression, ComponentVariable, StateExpression};
 use crate::TransitionSystems::{CompositionType, LocationID, LocationTree, TransitionSystemPtr};
-use std::slice::Iter;
 
-/// This function takes a [`QueryExpression`], the system recipe, and the transitionsystem -
-/// to define a state from the [`QueryExpression`] which has clocks and locations.
-/// The [`QueryExpression`] looks like this: `State(Vec<LocName>, Option<BoolExpression>)`.
+/// This function takes a [`StateExpression`], the system recipe, and the transitionsystem -
+/// to define a state from the [`StateExpression`] which has clocks and locations.
 /// `state_query` is the part of the query that describes the location and the clock constraints of the state.
 /// `machine` defines which operators is used to define the transistion system.
 /// `system` is the transition system.
 pub fn get_state(
-    state_query: &QueryExpression,
-    machine: &SystemRecipe,
+    state_query: &StateExpression,
+    recipe: &SystemRecipe,
     system: &TransitionSystemPtr,
 ) -> Result<State, String> {
-    if let QueryExpression::State(loc, clock) = state_query {
-        let mut locations: Vec<&str> = Vec::new();
+    // Check that there are no duplicated names in the system recipe components
+    let components = recipe.get_components();
+    for c1 in &components {
+        for c2 in &components {
+            if *c1 == *c2 {
+                continue;
+            }
+            if c1.name == c2.name && c1.special_id == c2.special_id {
+                return Err(format!(
+                    "Ambiguous component name: {}[{:?}] and {}[{:?}] are indistinguishable",
+                    c1.name, c1.special_id, c2.name, c2.special_id
+                ));
+            }
+        }
+    }
```

Comment:
Personally, I think this problem (finding identical components) is more suited to be solved by using iterators: 
```rust
if let Some((c1, c2)) = components
        .iter()
        .cartesian_product(components.iter())
        .filter(|(c1, c2)| ***c1 != ***c2)
        .find(|(c1, c2)| c1.name == c2.name && c1.special_id == c2.special_id)
    {
        return Err(format!(
            "Ambiguous component name: {}[{:?}] and {}[{:?}] are indistinguishable",
            c1.name, c1.special_id, c2.name, c2.special_id
        ));
    }
```
While I personally find this more readable, I can see if some might not. I'll let you decide what to do.
(They same for locations below)